torch/_inductor/codegen/cpp.py

```diff
@@ -216,53 +216,70 @@ def _print_Integer(self, expr):
         x, div, mod = expr.args
         x = self.paren(self.doprint(x))
-        div = self.paren(self.doprint(div))
+        if div != 1:
+            div = self.paren(self.doprint(div))
+            if expr.is_integer:
+                x = f"at::native::div_floor_integer({x}, {div})"
+            else:
+                x = f"at::native::div_floor_floating(static_cast<double>({x}), static_cast<double>({div}))"
         mod = self.paren(self.doprint(mod))
-        if div != "1":
-            x = f"({x} / {div})"
-        return f"static_cast<{INDEX_TYPE}>({x}) % static_cast<{INDEX_TYPE}>({mod})"
+        return f"static_cast<long>({x}) % static_cast<{INDEX_TYPE}>({mod})"
 
     def _print_FloorDiv(self, expr):
         x, div = expr.args
         x = self.paren(self.doprint(x))
         div = self.paren(self.doprint(div))
-        return f"std::floor({x} / {div})"
+        if expr.is_integer:
+            return f"at::native::div_floor_integer({x}, {div})"
+        return f"at::native::div_floor_floating(static_cast<double>({x}), static_cast<double>({div}))"
 
     def _print_floor(self, expr):
         assert len(expr.args) == 1
+        # integer expr should not hit this:
+        # s1 = sympy.Symbol("s1", integer=integer)
+        # expr = sympy.ceiling(s1)
+        # expr is alway s1.
+        assert expr.is_integer is not True
         return f"std::floor({self._print(expr.args[0])})"
 
     def _print_Pow(self, expr):
         # Uses float constants to perform FP div
         base, exp = expr.args
         base = self._print(base)
         if exp == 0.5:
-            return f"std::sqrt({base})"
+            r = f"std::sqrt({base})"
+            return f"static_cast<long>({r})" if expr.is_integer else r
         assert exp.is_integer
         exp = int(exp)
         if exp > 0:
-            return "*".join([self.paren(base)] * exp)
+            r = "*".join([self.paren(base)] * exp)
         elif exp < 0:
-            return "1.0/" + self.paren("*".join([self.paren(base)] * abs(exp)))
+            r = "1.0/" + self.paren("*".join([self.paren(base)] * abs(exp)))
         else:  # exp == 0
-            return "1"
+            r = "1.0"
+
+        return f"static_cast<long>({r})" if expr.is_integer else r
 
     def _print_Rational(self, expr):
         # Uses float constants to perform FP div
         if expr.q == 1:
-            return f"{expr.p}"
-        return f"{expr.p}.0/{expr.q}.0"
+            r = f"{expr.p}"
+        else:
+            r = f"{expr.p}.0/{expr.q}.0"
+        return f"static_cast<long>({r})" if expr.is_integer else r
 
     def _print_ceiling(self, expr):
         assert len(expr.args) == 1
+        # As _print_floor, integer exp should not hit this:
+        assert expr.is_integer is not True
         return f"std::ceil({self._print(expr.args[0])})"
 
 
 cexpr = CppPrinter().doprint
 
 
 def cexpr_index(index):
-    return f"static_cast<{INDEX_TYPE}>({cexpr(index)})"
+    return f"static_cast<long>({cexpr(index)})"
```

Comment:
We use `INDEX_TYPE` for the cpp type of indices. I would suggest we keep this, not hard-coding `long` here. Also, if we are sure the functions like `ModularIndexing` always returns an index, I would suggest we use `INDEX_TYPE` instead of `long` too.

Reply:
Ok, updated.